### PR TITLE
Ability to set dynamo db indexes

### PR DESCRIPTION
**Pull request by ferrants**
http://boto.cloudhackers.com/en/latest/ref/dynamodb2.html#boto.dynamodb2.fields.GlobalIncludeIndex

see `boto.dynamodb2.fields.*Index` classes

### CODE REVIEW

File: cloud/amazon/dynamodb_table.py [hunk 1]
```diff
@@ -222,16 +251,43 @@ def dynamo_table_exists(table):
...
+        if not check_mode:
+            for name, index in added_indexes.iteritems():
+                # raise Exception(str(index.schema()))
+                global_indexes_changed = table.create_global_secondary_index(global_index=index) or global_indexes_changed
+        else:
+            global_indexes_changed = True
+
+    if index_throughput_changes:
+        if not check_mode:
+            # todo: remove try once boto has https://github.com/boto/boto/pull/3447 fixed
```
Comment (ferrants) [inline cloud/amazon/dynamodb_table.py:281]: I left todos for future work based on working around a boto bug. Let me know if I should remove them
https://github.com/boto/boto/pull/3447

File: cloud/amazon/dynamodb_table.py [hunk 2]
```diff
@@ -242,6 +298,81 @@ def has_throughput_changed(table, new_throughput):
...
+    table_index_info = {index.name: index.schema() for index in table.global_indexes}
+    table_index_objects = {index.name: index for index in table.global_indexes}
+    set_index_info = {index.name: index.schema() for index in global_indexes}
+    set_index_objects = {index.name: index for index in global_indexes}
+
+    removed_indexes = {name: index for name, index in table_index_info.iteritems() if name not in set_index_info}
+    added_indexes = {name: set_index_objects[name] for name, index in set_index_info.iteritems() if name not in table_index_info}
+    # todo: uncomment once boto has https://github.com/boto/boto/pull/3447 fixed
+    # index_throughput_changes = {name: index.throughput for name, index in set_index_objects.iteritems() if name not in added_indexes and (index.throughput['read'] != str(table_index_objects[name].throughput['read']) or index.throughput['write'] != str(table_index_objects[name].throughput['write']))}
+    # todo: remove once boto has https://github.com/boto/boto/pull/3447 fixed
```
Comment (ferrants) [inline cloud/amazon/dynamodb_table.py:325]: todos here around that same bug

**Comment by robynbergeron** [comment]
Thanks @ferrants. @loia please review according to guidelines (http://docs.ansible.com/ansible/developing_modules.html#module-checklist) and comment with text 'shipit' or 'needs_revision' as appropriate.

**Comment by gregdek** [comment]
@loia This change is still pending your review; do you have time to take a look and comment? Please comment with text 'shipit' or 'needs_revision' as appropriate.

**Comment by ferrants** [comment]
all my testing with the module with these additions has checked out. been using for 2+ weeks without issue, indexes set correctly

**Comment by ferrants** [comment]
hey @gregdek, how could I get involved in being an ansible code reviewer?

**Comment by loia** [comment]
@ferrants Looks good overall, just a couple minor things.

The boto version in the `requirements` documentation needs to be updated.  From my testing, the min version it worked for me was `2.37.0`.  Ran into these issues with older boto versions:
- `global_indexes` not populated by `table.describe()`
- `table.update_global_secondary_index()` does not exist

There's a few PEP-8 issues which would be nice to fix up too e.g. line too long, too many blank lines, expected 2 blank lines.

needs_revision

**Comment by gregdek** [comment]
Thanks @ferrants for this PR. A maintainer of this module has asked for revisions to this PR. Please make the suggested revisions. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

**Comment by sivel** [comment]
```
============================================================================
./cloud/amazon/dynamodb_table.py
============================================================================
ERROR: version_added for new option (indexes) should be 2.1. Currently 0.0
```

**Comment by ferrants** [comment]
ready_for_review

File: cloud/amazon/dynamodb_table.py [hunk 3]
```diff
@@ -24,9 +24,8 @@
   - Returns the status of the specified table.
 version_added: "2.0"
 author: Alan Loi (@loia)
-version_added: "2.0"
```
Comment (ferrants) [inline cloud/amazon/dynamodb_table.py:27]: this was duplicated two lines above

File: cloud/amazon/dynamodb_table.py [hunk 4]
```diff
@@ -24,9 +24,8 @@
...
 requirements:
-  - "boto >= 2.13.2"
+  - "boto >= 2.37.0"
```
Comment (ferrants) [inline cloud/amazon/dynamodb_table.py:28]: set correct version @loia

File: cloud/amazon/dynamodb_table.py [hunk 5]
```diff
@@ -71,6 +70,15 @@
...
     default: 1
+  indexes:
+    description:
+      - list of dictionaries describing indexes to add to the table. global indexes can be updated. local indexes don't support updates or have throughput.
+      - required options: ['name', 'type', 'hash_key_name']
+      - valid types: ['all', 'global_all', 'global_include', 'global_keys_only', 'include', 'keys_only']
+      - other options: ['hash_key_type', 'range_key_name', 'range_key_type', 'includes', 'read_capacity', 'write_capacity']
+    required: false
+    default: []
+    version_added: "2.1"
```
Comment (ferrants) [inline cloud/amazon/dynamodb_table.py:81]: version added for new option here

**Comment by loia** [comment]
Thanks @ferrants, looks good.

shipit

**Comment by ferrants** [comment]
@gregdek can I get the `needs_revision` label pulled from this or have it merged? :smiley: thanks!

**Comment by gregdek** [comment]
Thanks again to @ferrants for this PR, and thanks @loia for reviewing. Marking for inclusion.

**Pull request merged**